slitlessutils/core/modules/extract/multi/matrix.py

```diff
@@ -638,8 +635,8 @@ def _lsmr(self, damp, **kwargs):
 
         (x, istop, itn, norm, arnorm, anorm, acond, xnorm) = linalg.lsmr(self.A, self.bi,
                                                                          damp=damp, **kwargs)
-        r1norm = rnorm
-        r2norm = np.sqrt(rnorm**2+(damp*xnorm)**2)
+        r1norm = arnorm
+        r2norm = np.sqrt(arnorm**2+(damp*xnorm)**2)
```

Comment:
@Russell-Ryan I took a guess here, since `rnorm` was not defined anywhere and `arnorm` was.

Reply:
Yes, you are correct.  I checked the documentations for LSQR and LSMR in scipy.

For the record (if someone else in the universe reads this).  

norm(r) = norm(b - Ax), and norm is "square root of the sum of squares". In the context of pyLINEAR, b is the array of pixel values (in e-/s), x is the array of spectral values (in CGS units), and A is the linear operator that transforms between them (in the simplest case, you can think of this just a very fancy and elaborate unit conversion step).  But in practice, this is a chi2-square problem and so (b-Ax) is the residual in detector units, and so norm(r)**2 is just the chi2 we all know and love.  

The catch to all that is, we're not actually solving a chi2 problem, we're solving a damped-chi2 problem (or in fancy mathy-language its: chi2 with Tikhonov regularization).  And in this type of problem, chi2 isn't really the relevant thing, but this other thing, which LSQR calls r2norm.  and if the damping term (I called it `damp` in matrix.py) is 0, then r2norm = rnorm = chi (not chi2). The LSQR algorithm requires knowing the norm(x) (ie. the square-root of the sum of squares of the inferred spectral elements), which is effectively a measure of the variation in the spectra (but here "variation" means with respect to the known solution, so it's better thought of as signal-to-noise or something). So, in Tikhonov regularization problems, this thing r2norm is the key metric, but that does not appear in the LSMR implementation.  And I wanted the API of `_lsmr` and `_lsqr` to be the same, so I had to compute r2norm in `_lsmr`.  If it wasn't clear, `_lsmr` and `_lsqr` are just private functions to call the scipy implementations of LSMR and LSQR in a consistent way.  

For long-term vision....  There are other methods (such as GSLSQR) to solve these types of integral problems (which are usually called "over-constrained sparse linear systems").  My plan for v2.0 of the code would be to fully support these other methods, but LSQR is the only one that I have any significant experience with. 

I think that was way more info than you wanted.  But I thought maybe this would be helpful to understanding *why* you had this problem in lsmr but not in lsqr and what these variables represent.  I expect to document these r1norm, r2norm, anorm, xnorm, arnorm a bit more.  It's all in scipy, but I always get frustrated when I have to look it up every time.